tensorflow_io/avro/20190307-avro-dataset.md

```diff
@@ -0,0 +1,299 @@
+- Transform primitive and nested Avro types into TensorFlow sparse and dense tensors
+- The creation of dimensions by (see also examples in Tables 2 and 3)
+  - Combining of several rows -- also called batching
+  - Combining several Avro fields to define indices and values for sparse tensors
+  - Combining and reshaping of elements of an Avro non-nested array into a multi-dimensional tensor using row-major-order
+  - Nestings of Avro typed arrays
+- Filling in scalars or tensors for missing values whenever the number of provided elements does not match the desired number of elements by using default values -- if no defaults or the wrong defaults are provided this results in a runtime error surfaced through an exception in python and erroneous status returned in c/c++
+- Full support for N-dimensional dense tensors and N-dimensional sparse tensors
+- A mapping of Avro primitive types to TensorFlow primitive types (see Table 1)
+- For Avro&#39;s null type the default shall be assumed if provided; otherwise fail
+- An API/syntax that follows the one in Spark/Avro tooling that allows to resolve
+  - Elements in an array
+  - Elements in a map
+  - Nested records
+  - Namespaces
+  - Branches in unions -- will resolve them but will use strict typing which will render them useless whenever an attribute has multiple types or whenever a null is encountered. Note, TensorFlow uses strict typing for it&#39;s tensors.
+- An API for filtering in arrays
+- The mapping of Avro enumerations to TensorFlow strings carrying the symbol name
+- The mapping of Avro fixed types to TensorFlow strings
+- The same operating systems as supported by TensorFlow -- Note, that this only accounts for the development of the avro dataset and not it&#39;s dependencies
+- Python 2.x and 3.x following TensorFlow&#39;s requirements
+- Thread safety; Note that TensorFlow&#39;s  dataset API may use multiple threads
+
+| **Avro primitive type** | **TensorFlow primitive type** |
+| --- | --- |
+| null: no value | If provided default is assumed otherwise fail |
+| boolean: a binary value | tf.bool |
+| int: 32-bit signed integer | tf.int32 |
+| long: 64-bit signed integer | tf.in64 |
+| float: single precision 32-bit IEEE floating point number | tf.float32 |
+| double: double precision 64-bit IEEE floating point number | tf.float64 |
+| fixed: a fixed length byte sequence | tf.string |
+| bytes: a sequence of 8-bit unsigned bytes | tf.string |
+| string: unicode character sequence | tf.string |
+| enum: enumeration | tf.string |
+
+**Table 1** lists the mapping of Avro primitive types to TensorFlow primitive types.
+
+The Avro dataset shall NOT support
+
+- --Type promotion or type coercion when mapping from Avro types to TensorFlow types, instead we use strict typing here
+- --Recursive type definitions in avro
+- --The compile and distribution of TensorFlow external tools that are necessary; e.g. libavro.so. It assumed that the user will compile and distribute this library separately.
```

Comment:
We could see what is the best way to build, either link externally or static link. I think I could help either case with bazel.

Reply:
Thanks. I'll attempt to convert the build from here (https://github.com/apache/avro/tree/master/lang/c) which would allows us to run through the tests as well. This is preferred.

Reply:
+1 on having avroc built as part of the tensorflow/io Bazel build. More portable and maintainable.